public/app/plugins/panel/geomap/utils/utils.ts

```diff
@@ -0,0 +1,70 @@
+import { SelectableValue } from '@grafana/data';
+import { DataFrame, GrafanaTheme2 } from '@grafana/data/src';
+import { getColorDimension, getScalarDimension, getScaledDimension, getTextDimension } from 'app/features/dimensions';
+import { getGrafanaDatasource } from 'app/plugins/datasource/grafana/datasource';
+
+import { defaultStyleConfig, StyleConfig, StyleConfigState, StyleDimensions } from '../style/types';
+
+export function getStyleDimension(
+  frame: DataFrame | undefined,
+  style: StyleConfigState,
+  theme: GrafanaTheme2,
+  customStyleConfig?: StyleConfig
+) {
+  const dims: StyleDimensions = {};
+  if (customStyleConfig && Object.keys(customStyleConfig).length) {
+    dims.color = getColorDimension(frame, customStyleConfig.color ?? defaultStyleConfig.color, theme);
+    dims.size = getScaledDimension(frame, customStyleConfig.size ?? defaultStyleConfig.size);
+    dims.rotation = getScalarDimension(frame, customStyleConfig.rotation ?? defaultStyleConfig.rotation);
+    if (customStyleConfig.text && (customStyleConfig.text.field || customStyleConfig.text.fixed)) {
+      dims.text = getTextDimension(frame, customStyleConfig.text!);
+    }
+  } else {
+    if (style.fields) {
+      if (style.fields.color) {
+        dims.color = getColorDimension(frame, style.config.color ?? defaultStyleConfig.color, theme);
+      }
+      if (style.fields.size) {
+        dims.size = getScaledDimension(frame, style.config.size ?? defaultStyleConfig.size);
+      }
+      if (style.fields.text) {
+        dims.text = getTextDimension(frame, style.config.text!);
+      }
+      if (style.fields.rotation) {
+        dims.rotation = getScalarDimension(frame, style.config.rotation ?? defaultStyleConfig.rotation);
+      }
+    }
+  }
+
+  return dims;
+}
+
+let publicGeoJSONFiles: Array<SelectableValue<string>> | undefined = undefined;
+
+export function getPublicGeoJSONFiles(): Array<SelectableValue<string>> {
+  if (!publicGeoJSONFiles) {
+    publicGeoJSONFiles = [];
+    initGeojsonFiles(); // async
+  }
+  return publicGeoJSONFiles;
+}
+
+// This will find all geojson files in the maps and gazetteer folders
+async function initGeojsonFiles() {
+  const ds = await getGrafanaDatasource();
+  for (let folder of ['maps', 'gazetteer']) {
+    ds.listFiles(folder).subscribe({
+      next: (frame) => {
+        frame.forEach((item) => {
+          if (item.name.endsWith('.geojson')) {
+            const value = `public/${folder}/${item.name}`;
+            publicGeoJSONFiles!.push({
+              value,
+              label: value,
+            });
+          }
+        });
+      },
+    });
+  }
+}
```

Comment:
This structure might be a little bit more readable
```suggestion
export async function getPublicGeoJSONFiles(): Array<SelectableValue<string>> {
  if (!publicGeoJSONFiles) {
    publicGeoJSONFiles = [];
    publicGeoJSONFiles = await initGeojsonFiles();
  }
  return publicGeoJSONFiles;
}

// This will find all geojson files in the maps and gazetteer folders
async function initGeojsonFiles() {
  const geoJSONFiles: Array<SelectableValue<string>> = []
  const ds = await getGrafanaDatasource();
  for (let folder of ['maps', 'gazetteer']) {
    ds.listFiles(folder).subscribe({
      next: (frame) => {
        frame.forEach((item) => {
          if (item.name.endsWith('.geojson')) {
            const value = `public/${folder}/${item.name}`;
            geoJSONFiles.push({
              value,
              label: value,
            });
          }
        });
      },
    });
  }
  return geoJSONFiles;
}
```

Reply:
I agree this would be better... but then it can only be called from async functions :(

As is it is called from a non-async function, but is updated frequently, so the proposed code is really a lazy loader (same array instance gets loaded)

Reply:
I think we could make the `registerOptionsUI` function in `geojsonLayer` an `async` function and then have access to the `await` keyword. Not sure though how our editor builder abstraction code handles async / if that breaks anything

I mean if it still works then I don't feel too strongly either way 🤷‍♂️